java-checks/src/test/files/checks/EqualsArgumentType.java

```diff
@@ -1,231 +1,224 @@
-
   @Override
-  public boolean equals(Object that) { // Noncompliant {{Add a type test to this method.}}
-    return that == that;
+  public boolean equals(Object that) { // Compliant, implicit call of a method in this
+    if (method(that)) {
+      return ((Object) that) == null;
+    }
   }
-
 }
 
-public class TestClassExplicitComparison5 {
-
-  @Override
-  public boolean equals(Object that) { // Noncompliant {{Add a type test to this method.}}
-    return that == "";
+public class TestMethodCall3 {
+  public boolean method(Object that) {
+    return that instanceof TestMethodCall;
   }
 
-}
-
-public class TestClassEqualsCall {
-
   @Override
-  public boolean equals(Object that) { // Compliant
-    return this.equals(that);
+  public boolean equals(Object that) { // Compliant, explicit call of a method in super
+    if (super.method(that)) {
+      return ((Object) that) == null;
+    }
   }
-
 }
 
-public class TestMethodCall {
-
-  public boolean method() {
-    return true;
+public class TestMethodCall4 {
+  public boolean method(Object that) {
+    return that instanceof TestMethodCall;
   }
 
   @Override
-  public boolean equals(Object that) { // Noncompliant {{Add a type test to this method.}}
-    return method();
+  public boolean equals(Object that) { // Compliant, explicit call of a method in this
+    if (this.method(that)) {
+      return ((Object) that) == null;
+    }
   }
-
 }
 
-public class TestClassInitializer {
+public class TestMethodCall5 {
+  Object[] objects;
 
-  @Override
-  public boolean equals(Object that) { // Compliant
-    boolean result = that.getClass() == Object.class && true;
-    return result;
+  public boolean method(Object that) {
+    return that instanceof TestMethodCall;
   }
 
-}
-
-public class TestClassNestedIf {
-
   @Override
-  public boolean equals(Object that) { // Compliant
-    if (true) {
-      return that.getClass() == Object.class;
+  public boolean equals(Object that) { // Noncompliant
+    if (objects[0].hashCode()) {
+      return ((Object) that) == null;
     }
   }
-
-}
-
-public class TestClassReturn {
-
-  @Override
-  public boolean equals(Object that) { // Compliant
-    return that.getClass() == Object.class && true;
-  }
-
 }
 
 public interface testInterface {
-
   public boolean equals(Object that); // Compliant
-
 }
 
 public class TestClass {
-
   public boolean method(Object that) { // Compliant, different name
+    return ((Object) that) == null;
   }
 
-  public boolean equals() { // Compliant, different signature
+  public Object equals() { // Compliant, different signature
+    return ((Object) that);
   }
 
   public boolean equals(String that) { // Compliant, different signature
+    return ((Object) that) == null;
   }
 
   @Override
-  public boolean equals(Object that) { // Noncompliant {{Add a type test to this method.}}
+  public boolean equals(Object that) { // Compliant
+    if (true) {
+      return (Object) this;
```

Comment:
Use another example of expression than a return statement... That example is hard to follow.